File: rotate-array.py
from typing import List

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k %= n  # 處理 k 大於 n 的情況

        def reverse(start, end):
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1

        reverse(0, n - 1)
        reverse(0, k - 1)
        reverse(k, n - 1)


s = Solution()
nums = [1,2,3,4,5,6,7]
s.rotate(nums, 3)
print(nums)

nums = [-1,-100,3,99]
s.rotate(nums, 2)
print(nums)

File: cpp/rotate-array.cpp
/* https://leetcode.com/discuss/27387/summary-of-c-solutions */
#include <iostream>
#include <iterator>
#include <vector>

using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        auto iter = nums.begin();
        for(; k %= n; n -= k, iter += k) {
            for(int i = 0; i < k; i++)
                swap(iter[i], iter[n - k + i]);
        }
    }
};

int main(int argc, const char* argv[])
{
    Solution s;
    vector<int> nums = {1, 2, 3, 4, 5, 6, 7};
    s.rotate(nums, 3);
    copy(nums.begin(), nums.end(), ostream_iterator<int>(std::cout, " "));
}
